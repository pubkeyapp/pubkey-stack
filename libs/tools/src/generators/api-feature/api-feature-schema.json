{
  "$schema": "http://json-schema.org/schema",
  "$id": "ApiFeature",
  "title": "",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "The name of the feature.",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use?"
    },
    "app": {
      "type": "string",
      "description": "The name of the application you are adding the feature to.",
      "default": "api"
    },
    "label": {
      "type": "string",
      "description": "The label field for the model. Generally 'name' or 'title'.",
      "default": "name"
    },
    "model": {
      "type": "string",
      "description": "The name of the model for this feature."
    },
    "crud": {
      "type": "string",
      "description": "Command separated list of actors to create CRUD for.",
      "default": "admin,user"
    },
    "skipDataAccess": {
      "type": "boolean",
      "description": "Do not create a data access library for this feature.",
      "default": false
    },
    "skipE2e": {
      "type": "boolean",
      "description": "Do not create an e2e test for this feature.",
      "default": false
    },
    "skipFeature": {
      "type": "boolean",
      "description": "Do not create a feature library for this feature.",
      "default": false
    },
    "skipSdk": {
      "type": "boolean",
      "description": "Do not create the SDK file for this feature.",
      "default": false
    },
    "skipUtil": {
      "type": "boolean",
      "description": "Do not create a util library for this feature.",
      "default": true
    }
  },
  "required": ["name"]
}
