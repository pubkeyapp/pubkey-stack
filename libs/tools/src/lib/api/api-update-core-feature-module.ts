import { Tree } from '@nx/devkit'
import { addNamedImport } from '../utils/ast/add-named-import'
import { updateSourceFile } from '../utils/ast/update-source-file'

export function apiUpdateCoreFeatureModule(
  tree: Tree,
  path: string,
  {
    featureClass,
    featurePackage,
  }: {
    featureClass: string
    featurePackage: string
  },
) {
  updateSourceFile(tree, path, (source) => {
    addNamedImport(source, featurePackage, featureClass)

    const importsVariable = source.getVariableStatement('imports')

    if (importsVariable) {
      // FIXME: there might be a better way to add this import to the array of imports
      source.insertText(importsVariable.getEnd() - 3, `, ${featureClass}`)
    }
    return source
  })
}
