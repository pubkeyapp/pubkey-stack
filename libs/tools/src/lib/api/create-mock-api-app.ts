import { getProjects, Tree } from '@nx/devkit'
import { applicationGenerator, libraryGenerator, serviceGenerator } from '@nx/nest'

export async function createMockApiApp(tree: Tree, app: string) {
  // Build the mock app and core libs
  await applicationGenerator(tree, {
    directory: `apps/${app}`,
    name: app,
    projectNameAndRootFormat: 'as-provided',
    skipFormat: true,
  })
  // Create the core data access lib
  await libraryGenerator(tree, {
    directory: `libs/${app}/core/data-access`,
    name: `${app}-core-data-access`,
    projectNameAndRootFormat: 'as-provided',
    skipFormat: true,
  })
  const p = getProjects(tree).get(`${app}-core-data-access`)
  // Create the core service
  await serviceGenerator(tree, {
    directory: `${p.sourceRoot}/lib`,
    name: `${app}-core`,
    nameAndDirectoryFormat: 'as-provided',
    skipFormat: true,
  })
  // Create the core feature lib
  await libraryGenerator(tree, {
    directory: `libs/${app}/core/feature`,
    name: `${app}-core-feature`,
    projectNameAndRootFormat: 'as-provided',
    skipFormat: true,
  })
  // Create the sdk lib
  await libraryGenerator(tree, {
    directory: `libs/${app}/sdk`,
    name: `sdk`,
    projectNameAndRootFormat: 'as-provided',
    skipFormat: true,
  })

  tree.write(
    'prisma/schema.prisma',
    `
  model User {
    id Int @id
  }
  model Company {
    id Int @id
    name String
    location String
    phone String?
  }
  `,
  )
}
