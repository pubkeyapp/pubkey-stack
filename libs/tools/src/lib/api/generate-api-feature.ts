import { getProjects, Tree } from '@nx/devkit'
import apiCrudGenerator from '../../generators/api-crud/api-crud-generator'

import { generateApiLib } from './generate-api-lib'
import { NormalizedApiFeatureSchema } from './normalized-api-feature-schema'

export async function generateApiFeature(tree: Tree, options: NormalizedApiFeatureSchema) {
  const projects = getProjects(tree)
  const [dataAccess, feature, util] = [
    `${options.app}-${options.model}-data-access`,
    `${options.app}-${options.model}-feature`,
    `${options.app}-${options.model}-util`,
  ].map((name) => projects.get(name))

  if (!options.skipDataAccess && !dataAccess) {
    await generateApiLib(tree, 'data-access', options)
  }
  if (!options.skipFeature && !feature) {
    await generateApiLib(tree, 'feature', options)
  }
  if (!options.skipUtil && !util) {
    await generateApiLib(tree, 'util', options)
  }
  if (!options.skipDataAccess && !options.skipFeature && options.crud?.length) {
    for (const actor of options.crud) {
      await apiCrudGenerator(tree, { actor, ...options })
    }
  }
}
