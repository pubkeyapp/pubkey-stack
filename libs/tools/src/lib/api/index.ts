export * from './normalize-api-feature-schema'
export * from './generate-api-feature'
