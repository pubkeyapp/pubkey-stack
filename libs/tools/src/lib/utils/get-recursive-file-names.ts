import { Tree } from '@nx/devkit'

export function getRecursiveFileNames({
  tree,
  path,
  exclude = [],
}: {
  tree: Tree
  path: string
  exclude?: string[]
}): string[] {
  const contents: string[] = []
  const dir = tree.children(path)
  dir
    .filter((file) => !exclude.includes(`${path}/${file}`))
    .forEach((file) => {
      if (tree.isFile(`${path}/${file}`)) {
        contents.push(`${path}/${file}`)
      } else {
        contents.push(
          ...getRecursiveFileNames({ tree, path: `${path}/${file}`, exclude }).sort((a, b) => a.localeCompare(b)),
        )
      }
    })

  return contents
}
