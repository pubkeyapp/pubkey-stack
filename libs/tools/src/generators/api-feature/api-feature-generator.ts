import { formatFiles, Tree } from '@nx/devkit'
import { generateApiFeature, normalizeApiFeatureSchema } from '../../lib/api'
import { ensureNxProjectExists } from '../../lib/utils/ensure-nx-project-exists'
import { ApiFeatureGeneratorSchema } from './api-feature-schema'

export async function apiFeatureGenerator(tree: Tree, rawOptions: ApiFeatureGeneratorSchema) {
  const options = normalizeApiFeatureSchema(tree, rawOptions)
  ensureNxProjectExists(tree, options.app)
  await generateApiFeature(tree, options)
  await formatFiles(tree)
}

export default apiFeatureGenerator
