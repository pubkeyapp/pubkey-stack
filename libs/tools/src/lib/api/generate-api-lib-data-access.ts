import { generateFiles, Tree } from '@nx/devkit'

import { replaceExports } from '../utils/ast/add-export'
import { ensureNxProjectExists } from '../utils/ensure-nx-project-exists'
import { getApiSubstitutions } from './get-api-substitutions'
import { NormalizedApiFeatureSchema } from './normalized-api-feature-schema'

export async function generateApiLibDataAccess(tree: Tree, options: NormalizedApiFeatureSchema) {
  const substitutions = getApiSubstitutions(options)

  const [dataAccess] = [`${options.app}-${options.model}-data-access`].map((project) =>
    ensureNxProjectExists(tree, project),
  )

  // Remove the generated data access module
  tree.delete(`${dataAccess.sourceRoot}/lib/${dataAccess.name}.module.ts`)
  // Generate the data access library
  generateFiles(tree, `${__dirname}/files/data-access`, dataAccess.sourceRoot, { ...substitutions })

  // Add the exports to the barrel file
  replaceExports(tree, `${dataAccess.sourceRoot}/index.ts`, [
    `./lib/${options.app}-${options.model}.data-access.module`,
    `./lib/${options.app}-${options.model}.service`,
  ])
}
