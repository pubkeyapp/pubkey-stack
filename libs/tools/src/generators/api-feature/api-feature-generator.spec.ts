import { readProjectConfiguration, Tree } from '@nx/devkit'
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing'
import { createMockApiApp } from '../../lib/api/create-mock-api-app'

import { getRecursiveFileNames } from '../../lib/utils/get-recursive-file-names'

import { apiFeatureGenerator } from './api-feature-generator'
import { ApiFeatureGeneratorSchema } from './api-feature-schema'

describe('api-feature generator', () => {
  let tree: Tree
  const options: ApiFeatureGeneratorSchema = { app: 'api', label: 'name', model: 'test' }

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace()
  })

  it('should generate the minimal feature libraries', async () => {
    await createMockApiApp(tree, options.app)

    // By default, we generate two libraries: data-access and feature
    const libs = ['data-access', 'feature']
    await apiFeatureGenerator(tree, options)

    libs.forEach((lib) => {
      const config = readProjectConfiguration(tree, `${options.app}-${options.model}-${lib}`)
      expect(config).toBeDefined()
    })

    const basePathDataAccess = `libs/${options.app}/${options.model}/data-access/src`
    const basePathFeature = `libs/${options.app}/${options.model}/feature/src`

    const sourceFilesDataAccess = getRecursiveFileNames({ tree, path: basePathDataAccess })
    const sourceFilesFeature = getRecursiveFileNames({ tree, path: basePathFeature })

    expect(sourceFilesDataAccess).toMatchInlineSnapshot(`
      [
        "libs/api/test/data-access/src/index.ts",
        "libs/api/test/data-access/src/lib/api-test.data-access.module.ts",
        "libs/api/test/data-access/src/lib/api-test.service.ts",
      ]
    `)
    expect(sourceFilesFeature).toMatchInlineSnapshot(`
      [
        "libs/api/test/feature/src/index.ts",
        "libs/api/test/feature/src/lib/api-test.feature.module.ts",
        "libs/api/test/feature/src/lib/api-test.resolver.ts",
      ]
    `)

    const files = [...sourceFilesDataAccess, ...sourceFilesFeature]

    files.forEach((file) => {
      expect(tree.exists(file)).toBeTruthy()
      expect(tree.read(file).toString()).toMatchSnapshot()
    })
  })

  it('should generate the feature libraries with crud for admin and user ', async () => {
    await createMockApiApp(tree, options.app)

    // By default, we generate two libraries: data-access and feature
    const libs = ['data-access', 'feature']
    await apiFeatureGenerator(tree, { ...options, crud: 'admin,user' })

    libs.forEach((lib) => {
      const config = readProjectConfiguration(tree, `${options.app}-${options.model}-${lib}`)
      expect(config).toBeDefined()
    })

    const basePathDataAccess = `libs/${options.app}/${options.model}/data-access/src`
    const basePathFeature = `libs/${options.app}/${options.model}/feature/src`

    const sourceFilesDataAccess = getRecursiveFileNames({ tree, path: basePathDataAccess })
    const sourceFilesFeature = getRecursiveFileNames({ tree, path: basePathFeature })

    expect(sourceFilesDataAccess).toMatchInlineSnapshot(`
      [
        "libs/api/test/data-access/src/index.ts",
        "libs/api/test/data-access/src/lib/api-test-data-admin.service.ts",
        "libs/api/test/data-access/src/lib/api-test-data-user.service.ts",
        "libs/api/test/data-access/src/lib/api-test-data.service.ts",
        "libs/api/test/data-access/src/lib/api-test.data-access.module.ts",
        "libs/api/test/data-access/src/lib/api-test.service.ts",
        "libs/api/test/data-access/src/lib/dto/test-admin-create.input.ts",
        "libs/api/test/data-access/src/lib/dto/test-admin-find-many.input.ts",
        "libs/api/test/data-access/src/lib/dto/test-admin-update.input.ts",
        "libs/api/test/data-access/src/lib/dto/test-user-create.input.ts",
        "libs/api/test/data-access/src/lib/dto/test-user-find-many.input.ts",
        "libs/api/test/data-access/src/lib/dto/test-user-update.input.ts",
        "libs/api/test/data-access/src/lib/entity/test.entity.ts",
        "libs/api/test/data-access/src/lib/helpers/get-test-where-admin.input.ts",
        "libs/api/test/data-access/src/lib/helpers/get-test-where-user.input.ts",
      ]
    `)
    expect(sourceFilesFeature).toMatchInlineSnapshot(`
      [
        "libs/api/test/feature/src/index.ts",
        "libs/api/test/feature/src/lib/api-test-admin.resolver.ts",
        "libs/api/test/feature/src/lib/api-test-user.resolver.ts",
        "libs/api/test/feature/src/lib/api-test.feature.module.ts",
        "libs/api/test/feature/src/lib/api-test.resolver.ts",
      ]
    `)

    const files = [...sourceFilesDataAccess, ...sourceFilesFeature]

    files.forEach((file) => {
      expect(tree.exists(file)).toBeTruthy()
      expect(tree.read(file).toString()).toMatchSnapshot()
    })
  })
  it('should generate the feature libraries with util lib', async () => {
    await createMockApiApp(tree, options.app)

    // By default, we generate two libraries: data-access and feature
    const libs = ['data-access', 'feature', 'util']
    await apiFeatureGenerator(tree, { ...options, skipUtil: false })

    libs.forEach((lib) => {
      const config = readProjectConfiguration(tree, `${options.app}-${options.model}-${lib}`)
      expect(config).toBeDefined()
    })
  })

  it('should generate the feature with different name', async () => {
    const testOptions = { ...options, model: 'company' }
    await createMockApiApp(tree, testOptions.app)

    // By default, we generate two libraries: data-access and feature
    const libs = ['data-access', 'feature', 'util']
    await apiFeatureGenerator(tree, { ...testOptions, skipUtil: false })

    libs.forEach((lib) => {
      const config = readProjectConfiguration(tree, `${testOptions.app}-${testOptions.model}-${lib}`)
      expect(config).toBeDefined()
    })
  })

  it('should generate the feature libraries with custom names', async () => {
    await createMockApiApp(tree, options.app)

    // By default, we generate two libraries: data-access and feature
    const libs = ['data-access', 'feature']
    const customName = 'custom'
    await apiFeatureGenerator(tree, { ...options, model: customName })

    libs.forEach((lib) => {
      const config = readProjectConfiguration(tree, `${options.app}-${customName}-${lib}`)
      expect(config).toBeDefined()
    })
  })
})
