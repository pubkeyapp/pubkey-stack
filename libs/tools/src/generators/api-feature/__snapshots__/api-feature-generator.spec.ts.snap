// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api-feature generator should generate the feature libraries with crud for admin and user  3`] = `
"export * from './lib/api-test.data-access.module';
export * from './lib/api-test.service';
export * from './lib/entity/test.entity';
export * from './lib/dto/test-admin-create.input';
export * from './lib/dto/test-admin-find-many.input';
export * from './lib/dto/test-admin-update.input';
export * from './lib/dto/test-user-create.input';
export * from './lib/dto/test-user-find-many.input';
export * from './lib/dto/test-user-update.input';
"
`;

exports[`api-feature generator should generate the feature libraries with crud for admin and user  4`] = `
"import { Injectable } from '@nestjs/common';
import { TestAdminCreateInput } from './dto/test-admin-create.input';
import { TestAdminFindManyInput } from './dto/test-admin-find-many.input';
import { TestAdminUpdateInput } from './dto/test-admin-update.input';
import { TestPaging } from './entity/test.entity';
import { getTestWhereAdminInput } from './helpers/get-test-where-admin.input';
import { ApiTestDataService } from './api-test-data.service';

@Injectable()
export class ApiTestDataAdminService {
  constructor(private readonly data: ApiTestDataService) {}

  async createTest(input: TestAdminCreateInput) {
    return this.data.create(input);
  }

  async deleteTest(testId: string) {
    return this.data.delete(testId);
  }

  async findManyTest(input: TestAdminFindManyInput): Promise<TestPaging> {
    return this.data.findMany({
      orderBy: { createdAt: 'desc' },
      where: getTestWhereAdminInput(input),
      limit: input.limit,
      page: input.page,
    });
  }

  async findOneTest(testId: string) {
    return this.data.findOne(testId);
  }

  async updateTest(testId: string, input: TestAdminUpdateInput) {
    return this.data.update(testId, input);
  }
}
"
`;

exports[`api-feature generator should generate the feature libraries with crud for admin and user  5`] = `
"import { Injectable } from '@nestjs/common';
import { TestUserCreateInput } from './dto/test-user-create.input';
import { TestUserFindManyInput } from './dto/test-user-find-many.input';
import { TestUserUpdateInput } from './dto/test-user-update.input';
import { TestPaging } from './entity/test.entity';
import { getTestWhereUserInput } from './helpers/get-test-where-user.input';
import { ApiTestDataService } from './api-test-data.service';

@Injectable()
export class ApiTestDataUserService {
  constructor(private readonly data: ApiTestDataService) {}

  async createTest(input: TestUserCreateInput) {
    return this.data.create(input);
  }

  async deleteTest(testId: string) {
    return this.data.delete(testId);
  }

  async findManyTest(input: TestUserFindManyInput): Promise<TestPaging> {
    return this.data.findMany({
      orderBy: { createdAt: 'desc' },
      where: getTestWhereUserInput(input),
      limit: input.limit,
      page: input.page,
    });
  }

  async findOneTest(testId: string) {
    return this.data.findOne(testId);
  }

  async updateTest(testId: string, input: TestUserUpdateInput) {
    return this.data.update(testId, input);
  }
}
"
`;

exports[`api-feature generator should generate the feature libraries with crud for admin and user  6`] = `
"import { Injectable } from '@nestjs/common';
import { Prisma } from '@prisma/client';
import { ApiCoreService, PagingInputFields } from '@proj/api-core-data-access';
import { TestPaging } from './entity/test.entity';

@Injectable()
export class ApiTestDataService {
  constructor(private readonly core: ApiCoreService) {}

  async create(input: Prisma.TestUncheckedCreateInput) {
    return this.core.data.test.create({ data: input });
  }

  async delete(testId: string) {
    await this.findOne(testId);
    const deleted = await this.core.data.test.delete({ where: { id: testId } });
    return !!deleted;
  }

  async findMany({
    limit = 10,
    page = 1,
    ...input
  }: Prisma.TestFindManyArgs & PagingInputFields): Promise<TestPaging> {
    return this.core.data.test
      .paginate(input)
      .withPages({ limit, page })
      .then(([data, meta]) => ({ data, meta }));
  }

  async findOne(testId: string) {
    const found = await this.core.data.test.findUnique({
      where: { id: testId },
    });
    if (!found) {
      throw new Error('Test not found');
    }
    return found;
  }

  async update(testId: string, input: Prisma.TestUpdateInput) {
    return this.core.data.test.update({ where: { id: testId }, data: input });
  }
}
"
`;

exports[`api-feature generator should generate the feature libraries with crud for admin and user  7`] = `
"import { Module } from '@nestjs/common';
import { ApiCoreDataAccessModule } from '@proj/api-core-data-access';
import { ApiTestService } from './api-test.service';
import { ApiTestDataService } from './api-test-data.service';
import { ApiTestDataAdminService } from './api-test-data-admin.service';
import { ApiTestDataUserService } from './api-test-data-user.service';

@Module({
  imports: [ApiCoreDataAccessModule],
  providers: [
    ApiTestService,
    ApiTestDataService,
    ApiTestDataAdminService,
    ApiTestDataUserService,
  ],
  exports: [ApiTestService],
})
export class ApiTestDataAccessModule {}
"
`;

exports[`api-feature generator should generate the feature libraries with crud for admin and user  8`] = `
"import { Injectable } from '@nestjs/common';
import { ApiTestDataService } from './api-test-data.service';
import { ApiTestDataAdminService } from './api-test-data-admin.service';
import { ApiTestDataUserService } from './api-test-data-user.service';

@Injectable()
export class ApiTestService {
  constructor(
    readonly data: ApiTestDataService,
    readonly admin: ApiTestDataAdminService,
    readonly user: ApiTestDataUserService
  ) {}
}
"
`;

exports[`api-feature generator should generate the feature libraries with crud for admin and user  9`] = `
"import { Field, InputType } from '@nestjs/graphql';

@InputType()
export class TestAdminCreateInput {
  @Field()
  name!: string;
}
"
`;

exports[`api-feature generator should generate the feature libraries with crud for admin and user  10`] = `
"import { Field, InputType } from '@nestjs/graphql';
import { PagingInput } from '@proj/api-core-data-access';

@InputType()
export class TestAdminFindManyInput extends PagingInput() {
  @Field({ nullable: true })
  search?: string;
}
"
`;

exports[`api-feature generator should generate the feature libraries with crud for admin and user  11`] = `
"import { Field, InputType } from '@nestjs/graphql';

@InputType()
export class TestAdminUpdateInput {
  @Field({ nullable: true })
  name?: string;
}
"
`;

exports[`api-feature generator should generate the feature libraries with crud for admin and user  12`] = `
"import { Field, InputType } from '@nestjs/graphql';

@InputType()
export class TestUserCreateInput {
  @Field()
  name!: string;
}
"
`;

exports[`api-feature generator should generate the feature libraries with crud for admin and user  13`] = `
"import { Field, InputType } from '@nestjs/graphql';
import { PagingInput } from '@proj/api-core-data-access';

@InputType()
export class TestUserFindManyInput extends PagingInput() {
  @Field({ nullable: true })
  search?: string;
}
"
`;

exports[`api-feature generator should generate the feature libraries with crud for admin and user  14`] = `
"import { Field, InputType } from '@nestjs/graphql';

@InputType()
export class TestUserUpdateInput {
  @Field({ nullable: true })
  name?: string;
}
"
`;

exports[`api-feature generator should generate the feature libraries with crud for admin and user  15`] = `
"import { Field, ObjectType } from '@nestjs/graphql';
import { PagingResponse } from '@proj/api-core-data-access';

@ObjectType()
export class Test {
  @Field()
  id!: string;
  @Field({ nullable: true })
  createdAt?: Date;
  @Field({ nullable: true })
  updatedAt?: Date;

  @Field()
  name!: string;
}

@ObjectType()
export class TestPaging extends PagingResponse<Test>(Test) {}
"
`;

exports[`api-feature generator should generate the feature libraries with crud for admin and user  16`] = `
"import { Prisma } from '@prisma/client';
import { TestAdminFindManyInput } from '../dto/test-admin-find-many.input';

export function getTestWhereAdminInput(
  input: TestAdminFindManyInput
): Prisma.TestWhereInput {
  const where: Prisma.TestWhereInput = {};

  if (input.search) {
    where.OR = [
      { id: { contains: input.search, mode: 'insensitive' } },
      { name: { contains: input.search, mode: 'insensitive' } },
    ];
  }

  return where;
}
"
`;

exports[`api-feature generator should generate the feature libraries with crud for admin and user  17`] = `
"import { Prisma } from '@prisma/client';
import { TestUserFindManyInput } from '../dto/test-user-find-many.input';

export function getTestWhereUserInput(
  input: TestUserFindManyInput
): Prisma.TestWhereInput {
  const where: Prisma.TestWhereInput = {};

  if (input.search) {
    where.OR = [
      { id: { contains: input.search, mode: 'insensitive' } },
      { name: { contains: input.search, mode: 'insensitive' } },
    ];
  }

  return where;
}
"
`;

exports[`api-feature generator should generate the feature libraries with crud for admin and user  18`] = `
"export * from './lib/api-test.feature.module';
"
`;

exports[`api-feature generator should generate the feature libraries with crud for admin and user  19`] = `
"import { Resolver } from '@nestjs/graphql';
import { ApiTestService } from '@proj/api-test-data-access';
import { ApiAuthGraphQLAdminGuard } from '@proj/api-auth-data-access';
import { Mutation, Query, Args } from '@nestjs/graphql';
import { UseGuards } from '@nestjs/common';
import {
  TestAdminCreateInput,
  TestAdminFindManyInput,
  Test,
  TestPaging,
  TestAdminUpdateInput,
} from '@proj/api-test-data-access';

@Resolver()
@UseGuards(ApiAuthGraphQLAdminGuard)
export class ApiTestAdminResolver {
  constructor(private readonly service: ApiTestService) {}

  @Mutation(() => Test, { nullable: true })
  adminCreateTest(@Args('input') input: TestAdminCreateInput) {
    return this.service.admin.createTest(input);
  }

  @Mutation(() => Boolean, { nullable: true })
  adminDeleteTest(@Args('testId') testId: string) {
    return this.service.admin.deleteTest(testId);
  }

  @Query(() => TestPaging)
  adminFindManyTest(@Args('input') input: TestAdminFindManyInput) {
    return this.service.admin.findManyTest(input);
  }

  @Query(() => Test, { nullable: true })
  adminFindOneTest(@Args('testId') testId: string) {
    return this.service.admin.findOneTest(testId);
  }

  @Mutation(() => Test, { nullable: true })
  adminUpdateTest(
    @Args('testId') testId: string,
    @Args('input') input: TestAdminUpdateInput
  ) {
    return this.service.admin.updateTest(testId, input);
  }
}
"
`;

exports[`api-feature generator should generate the feature libraries with crud for admin and user  20`] = `
"import { Resolver } from '@nestjs/graphql';
import { ApiTestService } from '@proj/api-test-data-access';
import { ApiAuthGraphQLUserGuard } from '@proj/api-auth-data-access';
import { Mutation, Query, Args } from '@nestjs/graphql';
import { UseGuards } from '@nestjs/common';
import {
  TestUserCreateInput,
  TestUserFindManyInput,
  Test,
  TestPaging,
  TestUserUpdateInput,
} from '@proj/api-test-data-access';

@Resolver()
@UseGuards(ApiAuthGraphQLUserGuard)
export class ApiTestUserResolver {
  constructor(private readonly service: ApiTestService) {}

  @Mutation(() => Test, { nullable: true })
  userCreateTest(@Args('input') input: TestUserCreateInput) {
    return this.service.user.createTest(input);
  }

  @Mutation(() => Boolean, { nullable: true })
  userDeleteTest(@Args('testId') testId: string) {
    return this.service.user.deleteTest(testId);
  }

  @Query(() => TestPaging)
  userFindManyTest(@Args('input') input: TestUserFindManyInput) {
    return this.service.user.findManyTest(input);
  }

  @Query(() => Test, { nullable: true })
  userFindOneTest(@Args('testId') testId: string) {
    return this.service.user.findOneTest(testId);
  }

  @Mutation(() => Test, { nullable: true })
  userUpdateTest(
    @Args('testId') testId: string,
    @Args('input') input: TestUserUpdateInput
  ) {
    return this.service.user.updateTest(testId, input);
  }
}
"
`;

exports[`api-feature generator should generate the feature libraries with crud for admin and user  21`] = `
"import { Module } from '@nestjs/common';
import { ApiTestDataAccessModule } from '@proj/api-test-data-access';
import { ApiTestResolver } from './api-test.resolver';
import { ApiTestAdminResolver } from './api-test-admin.resolver';
import { ApiTestUserResolver } from './api-test-user.resolver';

@Module({
  imports: [ApiTestDataAccessModule],
  providers: [ApiTestResolver, ApiTestAdminResolver, ApiTestUserResolver],
})
export class ApiTestFeatureModule {}
"
`;

exports[`api-feature generator should generate the feature libraries with crud for admin and user  22`] = `
"import { Resolver } from '@nestjs/graphql';
import { ApiTestService } from '@proj/api-test-data-access';
import { Test } from '@proj/api-test-data-access';

@Resolver(() => Test)
export class ApiTestResolver {
  constructor(private readonly service: ApiTestService) {}
}
"
`;

exports[`api-feature generator should generate the minimal feature libraries 3`] = `
"export * from './lib/api-test.data-access.module';
export * from './lib/api-test.service';
"
`;

exports[`api-feature generator should generate the minimal feature libraries 4`] = `
"import { Module } from '@nestjs/common';
import { ApiCoreDataAccessModule } from '@proj/api-core-data-access';
import { ApiTestService } from './api-test.service';

@Module({
  imports: [ApiCoreDataAccessModule],
  providers: [ApiTestService],
  exports: [ApiTestService],
})
export class ApiTestDataAccessModule {}
"
`;

exports[`api-feature generator should generate the minimal feature libraries 5`] = `
"import { Injectable } from '@nestjs/common';

@Injectable()
export class ApiTestService {
  constructor() {}
}
"
`;

exports[`api-feature generator should generate the minimal feature libraries 6`] = `
"export * from './lib/api-test.feature.module';
"
`;

exports[`api-feature generator should generate the minimal feature libraries 7`] = `
"import { Module } from '@nestjs/common';
import { ApiTestDataAccessModule } from '@proj/api-test-data-access';
import { ApiTestResolver } from './api-test.resolver';

@Module({
  imports: [ApiTestDataAccessModule],
  providers: [ApiTestResolver],
})
export class ApiTestFeatureModule {}
"
`;

exports[`api-feature generator should generate the minimal feature libraries 8`] = `
"import { Resolver } from '@nestjs/graphql';
import { ApiTestService } from '@proj/api-test-data-access';
import { Test } from '@proj/api-test-data-access';

@Resolver(() => Test)
export class ApiTestResolver {
  constructor(private readonly service: ApiTestService) {}
}
"
`;
